### PR TITLE
Implemented tracking of computation settings

### LINKED ISSUE #65: feat: add tracking of changed parameters when using comparison view

### LINKED ISSUE #86: fix(ui): area factor does not reset when resetting to historical settings
KnamLIJd6D

### LINKED ISSUE #87: fix: prevent app from crashing on pre-2005 election choices
Pre 2005 area factor is not used, because of that we hide the value. But we still check that the input is correct. The normal way of handling incorrect input is to return the default settings by validating them, but when the default settings are incorrect this causes infinite recursion.

### PR DESCRIPTION
**Pull request by Log234**
# Description
Keep track of changes to parameters when comparing different parameters.
It should be easy to see what parameters have been changed and what their original values were.

# Testing
Test that changes to all parameters, except year, are listed and correct. Also test that all listed parameters are reset correctly when resetting the parameters.

The parameters should be:
- Saved when the user clicks the "Lagre" button to save the computation
- Reset to previously saved settings when the user clicks "Gjenopprett"
- Reset to historical settings when the user changes the year or clicks "Tilbakestill instillinger"

## Setup
No change of configuration, run as normal.
## Expected
All changes should be listed and show what the original was. The listed parameters should also be removed when the parameters are reset.
## Issues closed
closes #65 
closes #86 
closes #87

### CODE REVIEW

**Comment by Log234** [comment]
Tracking is now implemented.
Currently, changes are displayed next to their respective parameters as it seemed natural.
It is however a design choice that should be evaluated and can be changed.

File: src/Layout/ComputationMenu/ComputationMenu.tsx [hunk 1]
```diff
@@ -307,6 +329,7 @@ export class ComputationMenu extends React.Component<ComputationMenuProps, {}> {
                         min={1}
                         max={5}
                         defaultValue={this.props.computationPayload.election.firstDivisor}
+                        originalValue={this.props.settingsPayload.comparison.firstDivisor}
```
Comment (simendag) [inline src/Layout/ComputationMenu/ComputationMenu.tsx]: Why was this done? Shouldn't display be controlled entirely by computation state?

File: src/common/SmartNumericInput/SmartNumericInput.tsx [hunk 2]
```diff
@@ -8,6 +8,7 @@ export interface SmartNumericInputProps {
     defaultValue: number;
     min: number;
     max: number;
+    originalValue?: string;
```
Comment (simendag) [inline src/common/SmartNumericInput/SmartNumericInput.tsx]: Why is this needed / used?

File: src/Layout/ComputationMenu/ComputationMenu.tsx [hunk 3]
```diff
@@ -317,6 +340,18 @@ export class ComputationMenu extends React.Component<ComputationMenuProps, {}> {
                         min={0}
                         max={15}
                         defaultValue={this.props.computationPayload.election.threshold}
+                        originalValue={this.props.settingsPayload.comparison.electionThreshold}
```
Comment (simendag) [inline src/Layout/ComputationMenu/ComputationMenu.tsx]: I suppose this goes under the SmartNumericInput change, but I am still unsure what originalValue actually does or how it benefits.

File: src/Layout/ComputationMenu/ConnectedComputationMenu.tsx [hunk 4]
```diff
@@ -133,6 +147,8 @@ const mapDispatchToProps = (
     saveComparison: () => {
         const saveComparisonAction = saveComparison();
         dispatch(saveComparisonAction);
+        const saveSettingsAction = saveSettings();
```
Comment (simendag) [inline src/Layout/ComputationMenu/ConnectedComputationMenu.tsx]: Why is it necessary to save/restore settings?

File: src/common/SmartNumericInput/SmartNumericInput.tsx [hunk 5]
```diff
@@ -49,6 +50,7 @@ export class SmartNumericInput extends React.Component<SmartNumericInputProps, {
                         />
                     )}
                 </div>
+                {this.props.originalValue !== this.props.value && <label>Orginalt: {this.props.originalValue}</label>}
```
Comment (simendag) [inline src/common/SmartNumericInput/SmartNumericInput.tsx]: This kind of calculation should probably be done in the scope of the calling function (render) rather than in the actual return.

File: src/Layout/ComputationMenu/AlgorithmSelect.tsx [hunk 6]
```diff
@@ -1,7 +1,9 @@
 import * as React from "react";
+import { getAlgorithmName } from "../../computation/logic";
 
 export interface AlgorithmSelectProps {
     algorithm: number;
+    defaultAlgorithm: number;
```
[the remark this answers is not in the document]
Comment (Log234) [reply src/Layout/ComputationMenu/AlgorithmSelect.tsx]: Sure. In other parts of the computation we are using the AlgorithmType. I believe the select component takes a number as the identifier, and that is why this was not originally implemented using Type. Should not be too difficult to fix.

File: src/Layout/ComputationMenu/ComputationMenu.tsx [hunk 7]
```diff
@@ -208,12 +230,10 @@ export class ComputationMenu extends React.Component<ComputationMenuProps, {}> {
             ...this.props.settingsPayload,
             areaFactor: stringValue,
         });
-        // Workaround for nested spread being a pain
-        const payload = this.props.computationPayload;
```
[the remark this answers is not in the document]
Comment (Log234) [reply src/Layout/ComputationMenu/ComputationMenu.tsx]: Your previous implementation was refering to an object within the computationPayload, Parameters. But that object is only supposed to reflect the historical parameters, which is why I updated the implementation to use a separate settings value in the base computationPayload, like we do for the other settings. I will go through the code once more and ensure there are no issues caused by this.

File: src/Layout/ComputationMenu/ComputationMenu.tsx [hunk 8]
```diff
@@ -249,8 +269,10 @@ export class ComputationMenu extends React.Component<ComputationMenuProps, {}> {
                     algorithm: getAlgorithmType(this.props.settingsPayload.algorithm),
                     firstDivisor: parseFloat(this.props.settingsPayload.firstDivisor),
                     electionThreshold: parseFloat(this.props.settingsPayload.electionThreshold),
+                    districtThreshold: parseFloat(this.props.settingsPayload.districtThreshold),
                     districtSeats: parseInt(this.props.settingsPayload.districtSeats),
                     levelingSeats: parseInt(this.props.settingsPayload.levelingSeats),
+                    areaFactor: parseInt(this.props.settingsPayload.areaFactor),
```
[the remark this answers is not in the document]
Comment (Log234) [reply src/Layout/ComputationMenu/ComputationMenu.tsx]: Nice spot, it is actually supposed to be a float, not int.

File: src/Layout/ComputationMenu/ComputationMenu.tsx (hunk 1 above)
Comment (Log234) [reply src/Layout/ComputationMenu/ComputationMenu.tsx]: No, the original value is keeping track of the settings we are comparing against. Hence it is not tied to the computation payload.

File: src/common/SmartNumericInput/SmartNumericInput.tsx (hunk 2 above)
Comment (Log234) [reply src/common/SmartNumericInput/SmartNumericInput.tsx]: It is needed to keep track of the value we are comparing against. But not all smart numeric inputs are election settings, which is why not all smart numeric inputs are supposed to have an original value.

File: src/Layout/ComputationMenu/ConnectedComputationMenu.tsx (hunk 4 above)
Comment (Log234) [reply src/Layout/ComputationMenu/ConnectedComputationMenu.tsx]: Because we want to track changes when we save/restore specific computations, not only changes against the historical settings.

File: src/common/SmartNumericInput/SmartNumericInput.tsx (hunk 5 above)
Comment (Log234) [reply src/common/SmartNumericInput/SmartNumericInput.tsx]: I agree, it would be tidier.

File: src/Layout/ComputationMenu/AlgorithmSelect.tsx [hunk 9]
```diff
@@ -27,6 +29,9 @@ export class AlgorithmSelect extends React.Component<AlgorithmSelectProps> {
                         </select>
                     </div>
                 </div>
+                {this.props.algorithm !== this.props.defaultAlgorithm && (
+                    <label>Orginalt: {getAlgorithmName(this.props.defaultAlgorithm)}</label>
+                )}
```
[the remark this answers is not in the document]
Comment (Log234) [reply src/Layout/ComputationMenu/AlgorithmSelect.tsx]: I agree. Will push a fix.
